### PR TITLE
[`Low-level-API`] Add docs about LLAPI

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

As discussed internally, this PR adds nice documentation and simple tests about the recent low-level API from #749 

cc @BenjaminBossan @pacman100

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -355,6 +355,42 @@ any GPU memory savings. Please refer issue [[FSDP] FSDP with CPU offload consume
 
 2. When using ZeRO3 with zero3_init_flag=True, if you find the gpu memory increase with training steps. we might need to update deepspeed after [deepspeed commit 42858a9891422abc](https://github.com/microsoft/DeepSpeed/commit/42858a9891422abcecaa12c1bd432d28d33eb0d4) . The related issue is [[BUG] Peft Training with Zero.Init() and Zero3 will increase GPU memory every forward step ](https://github.com/microsoft/DeepSpeed/issues/3002)
 
+## 🤗 PEFT as a utility library
+
+Inject trainable adapters on any `torch` model using `inject_adapter_in_model` method:
```
Comment (BenjaminBossan) [inline README.md:360]: Not sure exactly about the wording, but I think it's worth highlighting here that calling this function will only inject the adapters but make no further changes to the model. Otherwise, users may be confused why they should use this and not `get_peft_model`.

File: README.md [hunk 2]
```diff
@@ -355,6 +355,42 @@ any GPU memory savings. Please refer issue [[FSDP] FSDP with CPU offload consume
...
+lora_config = LoraConfig(
+    lora_alpha=16,
+    lora_dropout=0.1,
+    r=64,
+    bias="none",
+    target_modules=["linear"],
+)
+
+model = DummyModel()
+model = inject_adapter_in_model(lora_config, model, "default")
```
Comment (BenjaminBossan) [inline README.md:388]: Hmm, I wonder now why we did not have `adapter_name="default"` as a default argument? I think it would help here. If we don't want it, at least I would pass it as keyword argument in this example, not positional, to make it clear what the meaning of `"default"` is.

If we change the function to make the argument a default argument (ugh, so confusing, default vs `"default"`), the docs below also need to change a little bit ("that takes 3 arguments ...").

File: tests/test_low_level_api.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import unittest
+
+import torch
+
+from peft import LoraConfig, get_peft_model_state_dict, inject_adapter_in_model
+
+
+class DummyModel(torch.nn.Module):
```
Comment (BenjaminBossan) [inline tests/test_low_level_api.py:24]: Great that you added tests.

I think the tests could also be added to `test_custom_models.py`. The disadvantage of that change would be that as is, the tests are very clear and straightforward. The advantage would be that the tests in `test_custom_models.py` can be easily parametrized with different custom modules and configs, so the test coverage is better.

I would be okay if you want to keep it as is, it's just a suggestion.

File: docs/source/developer_guides/low_level_api.mdx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+## Pros and cons 
+
+When to use this API and when to not use it? Let's discuss in this section the pros and cons 
+
+Pros:
+- The model gets modified in-place, meaning the model will preserve all its original attributes and methods
+- Works for any torch module, and any modality (vision, text, multi-modal)
+
+Cons:
+- You need to manually writing saving and loading utility methods
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/low_level_api.mdx:102]:
```suggestion
- You need to manually write saving and loading utility methods
```

I think this could be confusing. I guess what you mean is stuff like `from_pretrained` etc. But people can still use the normal `torch.save` and `torch.load` that they already know. Saying they have to "manually" write the methods probably sounds worse than it actually is.

File: docs/source/developer_guides/low_level_api.mdx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+- You cannot use any of the utility method provided by `PeftModel` such as disabling adapters, merging adapters, etc.
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/low_level_api.mdx:103]: Maybe add a link to this section: https://huggingface.co/docs/peft/conceptual_guides/lora#utils-for-lora

Also, I took a look at some of the methods that are currently used for merging, unloading etc. and I think that with only a few changes, we can make them standalone functions (like `inject_adapter_in_model`) that don't require a `PeftModel` / `LoraModel` etc. At least for LoRA that should work.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for adding the docs and tests. They're very clear, I only have minor comments. Please take a look.

File: docs/source/developer_guides/low_level_api.mdx (hunk 4 above)
Comment (younesbelkada) [reply docs/source/developer_guides/low_level_api.mdx:102]: Hmm correct, let me rephrase this a bit then

File: docs/source/developer_guides/low_level_api.mdx (hunk 5 above)
Comment (younesbelkada) [reply docs/source/developer_guides/low_level_api.mdx:103]: Sounds great, we can do that in a follow up PR!

File: README.md (hunk 2 above)
Comment (younesbelkada) [reply README.md:388]: Sounds great, will address this change

File: tests/test_low_level_api.py (hunk 3 above)
Comment (younesbelkada) [reply tests/test_low_level_api.py:24]: I see ok thanks for explaining! I would say maybe let's keep it as it is since the test is aimed to only test the tiny snippet of the README and the docs so I want to keep it simple

**Comment by younesbelkada** [comment]
Thanks for the extensive review @BenjaminBossan ! Should have addressed the comments by now

**Review by BenjaminBossan — APPROVED** [review]
Fantastic, thanks for making the changes.
